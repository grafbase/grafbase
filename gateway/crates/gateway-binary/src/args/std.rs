use std::{
    fs,
    net::SocketAddr,
    path::{Path, PathBuf},
};

use anyhow::Context;
use ascii::AsciiString;
use clap::{ArgGroup, Parser};
use federated_server::{Config, GraphFetchMethod};
use grafbase_telemetry::{
    config::{OtlpExporterConfig, OtlpExporterGrpcConfig, OtlpExporterProtocol},
    otel::layer::BoxedLayer,
};
use graph_ref::GraphRef;
use tracing::Subscriber;
use tracing_subscriber::{registry::LookupSpan, Layer};

use super::{log::LogStyle, LogLevel};

#[derive(Debug, Parser)]
#[clap(
    group(
        ArgGroup::new("hybrid-or-airgapped")
            .required(true)
            .args(["graph_ref", "schema"])
    ),
    group(
        ArgGroup::new("graph-ref-with-access-token")
            .args(["graph_ref"])
            .requires("grafbase_access_token")
    )
)]
#[command(name = "Grafbase Gateway", version)]
/// Grafbase Gateway
pub struct Args {
    /// IP address on which the server will listen for incomming connections. Defaults to 127.0.0.1:5000.
    #[arg(short, long)]
    pub listen_address: Option<SocketAddr>,
    #[arg(short, long, help = GraphRef::ARG_DESCRIPTION, env = "GRAFBASE_GRAPH_REF", requires = "grafbase_access_token")]
    pub graph_ref: Option<GraphRef>,
    /// An access token to the Grafbase API. The scope must allow operations on the given account,
    /// and graph defined in the graph-ref argument.
    #[arg(env = "GRAFBASE_ACCESS_TOKEN", hide_env_values(true))]
    pub grafbase_access_token: Option<AsciiString>,
    /// Path to the TOML configuration file
    #[arg(long, short, env = "GRAFBASE_CONFIG_PATH")]
    pub config: Option<PathBuf>,
    /// Path to the schema SDL. If provided, the graph will be static and no connection is made
    /// to the Grafbase API.
    #[arg(long, short, env = "GRAFBASE_SCHEMA_PATH")]
    pub schema: Option<PathBuf>,
    /// Set the logging level
    #[arg(long = "log", env = "GRAFBASE_LOG")]
    pub log_level: Option<LogLevel>,
    /// Set the style of log output
    #[arg(long, env = "GRAFBASE_LOG_STYLE", default_value_t = LogStyle::Text)]
    log_style: LogStyle,
    /// If set, parts of the configuration will get reloaded when changed.
    #[arg(long, action)]
    hot_reload: bool,
}

impl super::Args for Args {
    /// The method of fetching a graph
    fn fetch_method(&self) -> anyhow::Result<GraphFetchMethod> {
        match self.graph_ref.as_ref() {
            Some(graph_ref) => Ok(GraphFetchMethod::FromApi {
                access_token: self
                    .grafbase_access_token
                    .clone()
                    .expect("present due to the arg group"),
                graph_name: graph_ref.graph().to_string(),
                branch: graph_ref.branch().map(ToString::to_string),
            }),
            None => {
                let federated_graph =
                    fs::read_to_string(self.schema.as_ref().expect("must exist if graph-ref is not defined"))
                        .context("could not read federated schema file")?;

                Ok(GraphFetchMethod::FromLocal {
                    federated_schema: federated_graph,
                })
            }
        }
    }

    fn config_path(&self) -> Option<&Path> {
        self.config.as_deref()
    }

    fn hot_reload(&self) -> bool {
        self.hot_reload
    }

    fn config(&self) -> anyhow::Result<Config> {
        let mut config = match self.config.as_ref() {
            Some(path) => {
                let config = fs::read_to_string(path).context("could not read config file")?;
                toml::from_str(&config)?
            }
            None => Config::default(),
        };

        if let Some(ref token) = self.grafbase_access_token {
            let mut telemetry = std::mem::take(&mut config.telemetry).unwrap_or_default();

            telemetry.grafbase = Some(OtlpExporterConfig {
                endpoint: "https://otel.grafbase.com".parse().unwrap(),
                enabled: true,
                protocol: OtlpExporterProtocol::Grpc,
                grpc: Some(OtlpExporterGrpcConfig {
                    tls: None,
                    headers: vec![(
                        AsciiString::from_ascii(b"authorization").context("Invalid auth header name")?,
                        AsciiString::from_ascii(format!("Bearer {token}")).context("Invalid access token")?,
                    )]
                    .into(),
                }),
                ..Default::default()
            });
            config.telemetry = Some(telemetry);
        }

        Ok(config)
    }

    fn log_format<S>(&self) -> BoxedLayer<S>
    where
        S: Subscriber + for<'span> LookupSpan<'span> + Send + Sync,
    {
        let layer = tracing_subscriber::fmt::layer();

        match self.log_style {
            // for interactive terminals we provide colored output
            LogStyle::Text if atty::is(atty::Stream::Stdout) => layer.with_ansi(true).with_target(false).boxed(),
            // for server logs, colors are off
            LogStyle::Text => layer.with_ansi(false).with_target(false).boxed(),
            LogStyle::Json => layer.json().boxed(),
        }
    }

    fn listen_address(&self) -> Option<SocketAddr> {
        self.listen_address
    }

    fn log_level(&self) -> Option<LogLevel> {
        self.log_level
    }
}
