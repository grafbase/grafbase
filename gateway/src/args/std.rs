use std::{
    fs,
    io::IsTerminal,
    net::SocketAddr,
    path::{Path, PathBuf},
};

use anyhow::Context;
use ascii::AsciiString;
use clap::{ArgGroup, Parser};
use federated_server::GraphFetchMethod;
use gateway_config::{BatchExportConfig, Config};
use grafbase_telemetry::config::{OtlpExporterConfig, OtlpExporterGrpcConfig, OtlpExporterProtocol};
use graph_ref::GraphRef;

use super::{log::LogStyle, LogLevel};

#[derive(Debug, Parser)]
#[clap(
    group(
        ArgGroup::new("hybrid-or-airgapped")
            .required(true)
            .args(["graph_ref", "schema"])
    ),
    group(
        ArgGroup::new("graph-ref-with-access-token")
            .args(["graph_ref"])
    )
)]
#[command(name = "Grafbase Gateway", version)]
/// Grafbase Gateway
pub struct Args {
    /// IP address on which the server will listen for incomming connections. Defaults to 127.0.0.1:5000.
    #[arg(short, long)]
    pub listen_address: Option<SocketAddr>,
    #[arg(short, long, help = GraphRef::ARG_DESCRIPTION, env = "GRAFBASE_GRAPH_REF")]
    pub graph_ref: Option<GraphRef>,
    /// Path to the TOML configuration file
    #[arg(long, short, env = "GRAFBASE_CONFIG_PATH")]
    pub config: Option<PathBuf>,
    /// Path to the schema SDL. If provided, the graph will be static and no connection is made
    /// to the Grafbase API.
    #[arg(long, short, env = "GRAFBASE_SCHEMA_PATH")]
    pub schema: Option<PathBuf>,
    /// Set the logging level, this applies to all spans, logs and trace events.
    ///
    /// Beware that *only* 'off', 'error', 'warn' and 'info' can be used safely in production. More
    /// verbose levels, such as 'debug', will include sensitive information like request variables, responses, etc.
    ///
    /// Possible values are: 'off', 'error', 'warn', 'info', 'debug', 'trace' or a custom string.
    /// In the last case, the string is passed on to [`tracing_subscriber::EnvFilter`] as is and is
    /// only meant for debugging purposes. No stability guarantee is made on the format.
    #[arg(long = "log", env = "GRAFBASE_LOG", default_value = "info")]
    pub log_level: String,
    /// Set the style of log output
    #[arg(long, env = "GRAFBASE_LOG_STYLE")]
    log_style: Option<LogStyle>,
    /// If set, parts of the configuration will get reloaded when changed.
    #[arg(long, action)]
    hot_reload: bool,
}

impl Args {
    pub fn grafbase_access_token(&self) -> Option<String> {
        std::env::var("GRAFBASE_ACCESS_TOKEN").ok()
    }
}

impl super::Args for Args {
    /// The method of fetching a graph
    fn fetch_method(&self) -> anyhow::Result<GraphFetchMethod> {
        match self.graph_ref.clone() {
            Some(graph_ref) => Ok(GraphFetchMethod::FromGraphRef {
                access_token: AsciiString::from_ascii(self.grafbase_access_token().ok_or_else(|| {
                    anyhow::format_err!(
                        "The GRAFBASE_ACCESS_TOKEN environment variable must be set when a graph_ref is provided"
                    )
                })?)?,
                graph_ref,
            }),
            None => {
                let federated_sdl =
                    fs::read_to_string(self.schema.as_ref().expect("must exist if graph-ref is not defined"))
                        .context("could not read federated schema file")?;

                Ok(GraphFetchMethod::FromSchema {
                    federated_sdl,
                    reload_signal: None,
                })
            }
        }
    }

    fn config_path(&self) -> Option<&Path> {
        self.config.as_deref()
    }

    fn hot_reload(&self) -> bool {
        self.hot_reload
    }

    fn config(&self) -> anyhow::Result<Config> {
        let mut config = match self.config.as_ref() {
            Some(path) => {
                let config = fs::read_to_string(path).context("could not read config file")?;
                toml::from_str(&config)?
            }
            None => Config::default(),
        };

        if let Some((token, graph_ref)) = self.grafbase_access_token().as_ref().zip(self.graph_ref.as_ref()) {
            config.telemetry.grafbase = Some(OtlpExporterConfig {
                endpoint: std::env::var("__GRAFBASE_OTEL_URL")
                    .unwrap_or("https://otel.grafbase.com".to_string())
                    .parse()
                    .unwrap(),
                enabled: true,
                protocol: OtlpExporterProtocol::Grpc,
                grpc: Some(OtlpExporterGrpcConfig {
                    tls: None,
                    headers: vec![
                        (
                            AsciiString::from_ascii(b"authorization").context("Invalid auth header name")?,
                            AsciiString::from_ascii(format!("Bearer {token}")).context("Invalid access token")?,
                        ),
                        (
                            AsciiString::from_ascii(b"grafbase-graph-ref").context("Invalid graph ref header name")?,
                            AsciiString::from_ascii(graph_ref.to_string().into_bytes()).context("Invalid graph ref")?,
                        ),
                    ]
                    .into(),
                }),
                batch_export: if let Some(seconds) = std::env::var("__GRAFBASE_OTEL_EXPORT_DELAY")
                    .ok()
                    .and_then(|v| v.parse::<usize>().ok())
                {
                    BatchExportConfig {
                        scheduled_delay: chrono::Duration::seconds(seconds as i64),
                        ..Default::default()
                    }
                } else {
                    Default::default()
                },
                ..Default::default()
            });
        }

        Ok(config)
    }

    fn log_style(&self) -> LogStyle {
        self.log_style.unwrap_or_else(|| {
            let log_level = self.log_level();
            if std::io::stdout().is_terminal() && (log_level.contains("debug") || log_level.contains("trace")) {
                LogStyle::Pretty
            } else {
                LogStyle::Text
            }
        })
    }

    fn listen_address(&self) -> Option<SocketAddr> {
        self.listen_address
    }

    fn log_level(&self) -> LogLevel<'_> {
        LogLevel(&self.log_level)
    }
}
