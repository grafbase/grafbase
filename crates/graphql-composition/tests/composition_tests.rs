use std::{fs, path::Path};

use graphql_composition::FederatedGraph;

fn run_test(test_path: &Path) -> anyhow::Result<()> {
    if cfg!(windows) {
        return Ok(()); // newlines
    }

    let test_description = fs::read_to_string(test_path)?;
    let subgraphs_dir = test_path.with_file_name("").join("subgraphs");
    let extensions_path = test_path.with_file_name("extensions.toml");

    if !subgraphs_dir.is_dir() {
        return Err(anyhow::anyhow!("{} is not a directory.", subgraphs_dir.display()));
    }

    let extensions: TestExtensions = fs::read_to_string(extensions_path)
        .ok()
        .map(|file| toml::from_str(&file).unwrap())
        .unwrap_or_default();

    let mut subgraphs_sdl = fs::read_dir(subgraphs_dir)?
        .filter_map(Result::ok)
        .filter(|file| file.file_name() != ".gitignore")
        .map(|file| fs::read_to_string(file.path()).map(|contents| (contents, file.path())))
        .collect::<Result<Vec<_>, _>>()?;

    // [fs::read_dir()] doesn't guarantee ordering. Sort to make tests deterministic
    // (inconsistencies observed in CI).
    subgraphs_sdl.sort_by_key(|(_, path)| path.file_name().unwrap().to_owned());

    let mut subgraphs = graphql_composition::Subgraphs::default();

    for (sdl, path) in subgraphs_sdl {
        let name = path.file_stem().unwrap().to_str().unwrap().replace('_', "-");

        subgraphs
            .ingest_str(&sdl, &name, Some(&format!("http://example.com/{name}")))
            .map_err(|err| anyhow::anyhow!("Error parsing {}: \n{err:#}", path.display()))?;
    }

    subgraphs.ingest_loaded_extensions(
        extensions
            .extensions
            .into_iter()
            .map(|extension| graphql_composition::LoadedExtension::new(extension.url, extension.name)),
    );

    let (federated_sdl, api_sdl) = match graphql_composition::compose(&subgraphs).into_result() {
        Ok(federated_graph) => (
            graphql_federated_graph::render_federated_sdl(&federated_graph).expect("rendering federated SDL"),
            Some(graphql_federated_graph::render_api_sdl(&federated_graph)),
        ),
        Err(diagnostics) => (
            format!(
                "{}\n",
                diagnostics
                    .iter_messages()
                    .map(|msg| format!("# {}", msg.lines().collect::<Vec<_>>().join("\\n")))
                    .collect::<Vec<_>>()
                    .join("\n"),
            ),
            None,
        ),
    };

    let test_description = Some(test_description.as_str().trim())
        .filter(|desc| !desc.is_empty())
        .unwrap_or("Federated SDL");

    insta::assert_snapshot!("federated.graphql", federated_sdl, test_description);

    if let Some(actual_api_sdl) = api_sdl {
        insta::assert_snapshot!("api.graphql", actual_api_sdl);
    }

    test_sdl_roundtrip(&federated_sdl)
}

fn test_sdl_roundtrip(sdl: &str) -> anyhow::Result<()> {
    // Exclude tests with an empty schema. This is the case for composition error tests.
    if sdl.lines().all(|line| line.is_empty() || line.starts_with('#')) {
        return Ok(());
    }

    let roundtripped = graphql_federated_graph::render_federated_sdl(
        &FederatedGraph::from_sdl(sdl).map_err(|err| anyhow::anyhow!("Error ingesting SDL: {err}\n\nSDL:\n{sdl}"))?,
    )?;

    if roundtripped == sdl {
        return Ok(());
    }

    pretty_assertions::assert_eq!(sdl, roundtripped, "SDL roundtrip failed");
    Ok(())
}

#[test]
fn composition_tests() {
    insta::glob!("composition/**/test.md", |test_path| {
        let snapshot_path = test_path.parent().unwrap();
        insta::with_settings!({
            snapshot_path => snapshot_path.to_str().unwrap(),
            prepend_module_to_snapshot => false,
            snapshot_suffix => "",
        }, {
            run_test(test_path).unwrap();
        });
    });
}

#[derive(Debug, serde::Deserialize, Default)]
struct TestExtensions {
    extensions: Vec<TestExtension>,
}

#[derive(Debug, serde::Deserialize)]
struct TestExtension {
    url: String,
    name: String,
}
