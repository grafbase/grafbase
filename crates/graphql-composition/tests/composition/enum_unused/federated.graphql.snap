---
source: crates/graphql-composition/tests/composition_tests.rs
expression: Federated SDL
input_file: crates/graphql-composition/tests/composition/enum_unused/test.md
---
directive @join__unionMember(graph: join__Graph!, member: String!) on UNION

directive @join__implements(graph: join__Graph!, interface: String!) on OBJECT | INTERFACE

directive @join__graph(name: String!, url: String) on ENUM_VALUE

directive @join__field(graph: join__Graph, requires: join__FieldSet, provides: join__FieldSet) on FIELD_DEFINITION

directive @join__type(graph: join__Graph, key: join__FieldSet, resolvable: Boolean = true) on OBJECT | INTERFACE

directive @join__owner(graph: join__Graph!) on OBJECT

scalar join__FieldSet

type User
    @join__type(graph: THESCHEMA)
{
    id: ID!
    name: String!
}

type Query
{
    user: User @join__field(graph: THESCHEMA)
}

enum UserType
    @join__type(graph: THESCHEMA)
{
    ADMIN
    REGULAR
}

enum join__Graph
{
    THESCHEMA @join__graph(name: "theschema", url: "http://example.com/theschema")
}
